go/vt/vttablet/tabletmanager/vreplication/external_connector.go

```diff
@@ -0,0 +1,177 @@
+)
+
+// vstreamerClient exposes the core interface of a vstreamer
+type vstreamerClient interface {
+	Open(context.Context) error
+	Close(context.Context) error
+
+	// VStream streams VReplication events based on the specified filter.
+	VStream(ctx context.Context, startPos string, filter *binlogdatapb.Filter, send func([]*binlogdatapb.VEvent) error) error
+
+	// VStreamRows streams rows of a table from the specified starting point.
+	VStreamRows(ctx context.Context, query string, lastpk *querypb.QueryResult, send func(*binlogdatapb.VStreamRowsResponse) error) error
+}
+
+type externalConnector struct {
+	mu         sync.Mutex
+	dbconfigs  map[string]*dbconfigs.DBConfigs
+	connectors map[string]*mysqlConnector
+}
+
+func newExternalConnector(dbcfgs map[string]*dbconfigs.DBConfigs) *externalConnector {
+	return &externalConnector{
+		dbconfigs:  dbcfgs,
+		connectors: make(map[string]*mysqlConnector),
+	}
+}
+
+func (ec *externalConnector) Close() {
+	for _, c := range ec.connectors {
+		c.shutdown()
+	}
+	ec.connectors = make(map[string]*mysqlConnector)
+}
+
+func (ec *externalConnector) Get(name string) (*mysqlConnector, error) {
+	ec.mu.Lock()
+	defer ec.mu.Unlock()
+	if c, ok := ec.connectors[name]; ok {
+		return c, nil
+	}
+
+	// Construct
+	config := tabletenv.NewDefaultConfig()
+	config.DB = ec.dbconfigs[name]
+	if config.DB == nil {
+		return nil, vterrors.Errorf(vtrpcpb.Code_NOT_FOUND, "external mysqlConnector %v not found", name)
+	}
+	c := &mysqlConnector{}
+	c.env = tabletenv.NewEnv(config, name)
+	c.se = schema.NewEngine(c.env)
+	c.vstreamer = vstreamer.NewEngine(c.env, nil, c.se)
+	c.se.InitDBConfig(c.env.Config().DB.DbaWithDB())
+
+	// Open
+	if err := c.se.Open(); err != nil {
+		return nil, vterrors.Wrapf(err, "external mysqlConnector: %v", name)
+	}
+	c.vstreamer.Open("", "")
+
+	// Register
+	ec.connectors[name] = c
+	return c, nil
+}
+
+//-----------------------------------------------------------
+
+type mysqlConnector struct {
+	env       tabletenv.Env
+	se        *schema.Engine
+	vstreamer *vstreamer.Engine
+}
+
+func (c *mysqlConnector) shutdown() {
+	c.vstreamer.Close()
+	c.se.Close()
+}
+
+func (c *mysqlConnector) Open(ctx context.Context) error {
+	return nil
+}
+
+func (c *mysqlConnector) Close(ctx context.Context) error {
+	return nil
+}
+
+func (c *mysqlConnector) VStream(ctx context.Context, startPos string, filter *binlogdatapb.Filter, send func([]*binlogdatapb.VEvent) error) error {
+	return c.vstreamer.Stream(ctx, startPos, filter, send)
+}
+
+func (c *mysqlConnector) VStreamRows(ctx context.Context, query string, lastpk *querypb.QueryResult, send func(*binlogdatapb.VStreamRowsResponse) error) error {
+	var row []sqltypes.Value
+	if lastpk != nil {
+		r := sqltypes.Proto3ToResult(lastpk)
+		if len(r.Rows) != 1 {
+			return vterrors.Errorf(vtrpcpb.Code_INVALID_ARGUMENT, "unexpected lastpk input: %v", lastpk)
+		}
+		row = r.Rows[0]
+	}
+	return c.vstreamer.StreamRows(ctx, query, row, send)
```

Comment:
Do we need to wait for lastpk to be nil and can it be done as a moving window thing. Basically is less than x rows are left in buffer, stream more. Is is worth thinking in these lines?

Reply:
The vreplication engine has a 1 hour timer during which it fetches as much as it can, and then resumes from where it left off. This keeps going till the whole table is scanned.